pallets/identity/src/keys.rs

```diff
@@ -244,6 +244,84 @@ impl<T: Config> Module<T> {
+            |data, target_did| {
+                let perms = extract_auth!(data, DirectRotatePrimaryKey(p));
+                // Ensure length limits are enforced in `perms`.
+                Self::ensure_perms_length_limited(&perms)?;
+
+                // Accept authorization from CDD service provider.
+                if Self::cdd_auth_for_primary_key_rotation() {
+                    let auth_id = optional_cdd_auth_id
+                        .ok_or_else(|| Error::<T>::InvalidAuthorizationFromCddProvider)?;
+
+                    let signer = Signatory::Account(new_primary_key.clone());
+                    Self::accept_auth_with(&signer, auth_id, |data, auth_by| {
+                        let attestation_for_did = extract_auth!(data, AttestPrimaryKeyRotation(a));
+                        // Attestor must be a CDD service provider.
+                        ensure!(
+                            T::CddServiceProviders::is_member(&auth_by),
+                            Error::<T>::NotCddProviderAttestation
+                        );
+                        // Ensure authorizations are for the same DID.
+                        ensure!(
+                            target_did == attestation_for_did,
+                            Error::<T>::AuthorizationsNotForSameDids
+                        );
+                        Ok(())
+                    })?;
+                }
+
+                DidRecords::<T>::mutate(target_did, |record| {
+                    // Ensure the new primary key is a secondary key.
+                    ensure!(
+                        record
+                            .secondary_keys
+                            .iter()
+                            .any(|si| si.signer == new_primary_key_signer),
+                        Error::<T>::NotASigner
+                    );
+                    let old_primary_key = record.primary_key.clone();
+
+                    // Remove old secondary key
+                    let removed_signers = vec![new_primary_key_signer.clone()];
+                    record.remove_secondary_keys(&removed_signers);
+                    Self::deposit_event(RawEvent::SecondaryKeysRemoved(
+                        target_did,
+                        removed_signers,
+                    ));
+
+                    // Ensure that it is safe to unlink the keys from the did.
+                    Self::ensure_key_unlinkable_from_did(&old_primary_key)?;
+                    Self::ensure_key_unlinkable_from_did(&new_primary_key)?;
+
+                    // Replace primary key of the owner that initiated key rotation.
+                    record.primary_key = new_primary_key.clone();
+                    Self::deposit_event(RawEvent::PrimaryKeyUpdated(
+                        target_did,
+                        old_primary_key.clone(),
+                        new_primary_key,
+                    ));
+
+                    Self::unsafe_join_identity(target_did, perms, old_primary_key);
```

Comment:
I recommend copying the add secondary key logic from `unsafe_join_identity` into here, since that also mutates the same `DidRecords` value.  Just need the `did_records.add_secondary_keys` call and the `SecondaryKeysAdded` event.  The `link_account_key_to_did` isn't needed because the old primary key is already linked.

It might be better to add a new event type for this action instead of emitting 3 different events.